### PR TITLE
magento/magento2:28748 : Virtual product row total remove from packag…

### PR DESCRIPTION
**Pull request by sunilit42**
fix issue :- https://github.com/magento/magento2/issues/28748

as descibe @NicholasRoge and thank you per decribe it

I debug based on NicholasRoge descibe and i added that columns and that is needed and based on that it is passing right value into package_physical_value

let me know if you need any screenshot or something

### CODE REVIEW

**Comment by gabrieldagama** [comment]
#dmcdindia2020

**Comment by sunilit42** [comment]
@magento run all tests

**Comment by aleron75** [comment]
@magento run Functional Tests EE

**Review by aleron75 — COMMENTED** [review]
Hi @sunilit42, thank you for your contribution.

Do you think it's possible for you to extend the existing Magento unit test (`\Magento\Quote\Test\Unit\Model\Quote\AddressTest`) to validate your fix?

Tests should fail on the mainline and pass with this fix.

Since this PR changes some core table structure, I have to put a medium risk label, the core team will decide whether some more regression testing needs to be done.

The functional tests (EE + B2B) that fail could not be related to this PR, I restarted EE to see if something changes.

**Comment by sunilit42** [comment]
Hello @aleron75 

i m not sure why i need to change into test case

already code is there, there is issue only that column is missing into db

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @sunilit42, thanks for your contributions. From my tests locally, just creating the column on the DB does not fix the problem, can you review it again?

Also, as described by @aleron75 we need to cover this PR changes with automated tests, the fact that we don't have failing tests without your fix means that this is not covered, and that has to be added in order to process this PR.

Thanks!

**Comment by sunilit42** [comment]
Hello,

https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Quote/Model/Quote/Address.php#L1050
on line no : $packagePhysicalValue = $item ? $item->getBaseRowTotal() : $baseSubtotal - $this->getBaseVirtualAmount();

And getBaseVirtualAmount it is assigned from
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Quote/Model/Quote/Address/Total/Subtotal.php#L69

**Comment by gabrieldagama** [comment]
Hello @sunilit42, did you test it locally and got `$this->getBaseVirtualAmount();` returning anything besides 0? From my tests data is not saved on the DB, it may be related to the fact that the data is being set on `Magento\Quote\Model\Quote\Address\Total` instead of the `Address` itself.

Also, `Magento\Quote\Model\Quote\Address\Total` doesn't have the `setBaseVirtualAmount` or `setVirtualAmount`, and we should implement them in a backward-compatible way.

Please don't forget to add test coverage as well, thank you!

**Comment by sunilit42** [comment]
@gabrieldagama 

https://i.imgur.com/D08kzRl.png : one simple product and one virtual product
https://i.imgur.com/5cJDkpP.png : in db you can see that value save and i did not any thing related to code beside added column into db

**Comment by engcom-Charlie** [comment]
Hello, @sunilit42 please implement `setBaseVirtualAmount()` and `setVirtualAmount()` on the `Total` object and cover your changes by some kind of autotest.
Thank you.

**Comment by engcom-Charlie** [comment]
Hello, @sunilit42 please look at https://github.com/magento/magento2/pull/30434#issuecomment-717996062. Otherwise, we can't proceed with your PR.
Thank you.

**Comment by sunilit42** [comment]
Hello @engcom-Charlie 

https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Quote/Test/Unit/Model/Quote/Address/Total/SubtotalTest.php#L162

->addMethods(['setVirtualAmount', 'setBaseVirtualAmount']) added that method implement if i will add into class then i need to remove from this line

**Comment by engcom-Charlie** [comment]
@sunilit42 it's no problem.

**Comment by engcom-Charlie** [comment]
Hello, @sunilit42 please look at #30434 (comment). Otherwise, we can't proceed with your PR.
Thank you.

**Comment by engcom-Charlie** [comment]
Hello, @sunilit42 please look at https://github.com/magento/magento2/pull/30434#issuecomment-717996062. Otherwise, we can't proceed with your PR.
Thank you.

**Comment by engcom-Charlie** [comment]
@sunilit42 please update your PR accordingly to https://github.com/magento/magento2/pull/30434#issuecomment-717996062.
Thank you.

**Comment by engcom-Charlie** [comment]
@sunilit42 I'll continue with your PR.

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by sunilit42** [comment]
@engcom-Charlie Thanks for your help !!!

File: app/code/Magento/Quote/Model/Quote/Address/Total.php
Comment (gabrieldagama) [inline app/code/Magento/Quote/Model/Quote/Address/Total.php:81]: I believe it makes sense to implement getters for such values as well.

File: app/code/Magento/Quote/Model/Quote/Address/Total.php
Comment (gabrieldagama) [inline app/code/Magento/Quote/Model/Quote/Address/Total.php:104]: I believe it makes sense to implement getters for such values as well.

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @engcom-Charlie,

Thanks for helping, can you have a look at my comments below?

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by aleron75** [comment]
@magento give me 2.4-develop instance

**Comment by aleron75** [comment]
@magento give me 2.4-develop instance

**Review by aleron75 — COMMENTED** [review]
Checking the code submitted with current PR, we (cc @danielrussob) see that it fixes the calculation of `$packagePhysicalValue`. 

Shouldn't also the `$packageWithDiscount` variable take virtual product value into account, subtracting it? Everything that is "virtual" should not be considered in "package" value, isn't it?

Otherwise, depending on how carrier total is calculated, the fix could not work in case of a discount applied to a mixed cart (physical + virtual product).

I'm not sure this bug is related to this PR or should be addressed as a new PR.

cc @engcom-Charlie

**Comment by danielrussob** [comment]
> Checking the code submitted with current PR, we (cc @danielrussob) see that it fixes the calculation of `$packagePhysicalValue`.
> 
> Shouldn't also the `$packageWithDiscount` variable take virtual product value into account, subtracting it? Everything that is "virtual" should not be considered in "package" value, isn't it?
> 
> Otherwise, depending on how carrier total is calculated, the fix could not work in case of a discount applied to a mixed cart (physical + virtual product).
> 
> I'm not sure this bug is related to this PR or should be addressed as a new PR.
> 
> cc @engcom-Charlie

Hello @sunilit42,
in addition to what was said by @aleron75 , please check:

`
\Magento\Fedex\Model\Carrier::setRequest:390
$r->setValue($request->getPackagePhysicalValue());
$r->setValueWithDiscount($request->getPackageValueWithDiscount());
`

we (cc: @aleron75 ) don't know if fedex use value_with_discount or value (or the lowest) in API

Furthermore, double check also:

`
\Magento\OfflineShipping\Model\Carrier\Freeshipping::isFreeShippingRequired:75
$minSubtotal = $request->getPackageValueWithDiscount(); <---it should contains physical + virtual price
`

and:

`
\Magento\Dhl\Model\Carrier::setRequest:488
->setValueWithDiscount($request->getPackageValueWithDiscount()) <---it should contains physical + virtual price
`

**Comment by engcom-Charlie** [comment]
Hello @sunilit42, could you please look at https://github.com/magento/magento2/pull/30434#issuecomment-732392265?
Thank you.

**Comment by engcom-Charlie** [comment]
Hi @aleron75 @danielrussob I can confirm that here
https://github.com/magento/magento2/blob/8c95f0780e8f9bb98987611a5bc0ecfe19bd9f15/app/code/Magento/Quote/Model/Quote/Address.php#L1042
we have in `$packageWithDiscount` also virtual value and as a result 
here
```
\Magento\Dhl\Model\Carrier::setRequest:488
->setValueWithDiscount($request->getPackageValueWithDiscount())
``` 
and here
```
\Magento\OfflineShipping\Model\Carrier\Freeshipping::isFreeShippingRequired:75
$minSubtotal = $request->getPackageValueWithDiscount();
```
we have also the virtual amount value.

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Review by danielrussob — APPROVED** [review]
The failure of "Semantic Version Checker" is generated only by:

Magento\Quote\Model\Quote\Address\Total::setVirtualAmount
/app/code/Magento/Quote/Model/Quote/Address/Total.php:81

Magento\Quote\Model\Quote\Address\Total::getVirtualAmount
/app/code/Magento/Quote/Model/Quote/Address/Total.php:93

Magento\Quote\Model\Quote\Address\Total::setBaseVirtualAmount
/app/code/Magento/Quote/Model/Quote/Address/Total.php:104

Magento\Quote\Model\Quote\Address\Total::getBaseVirtualAmount
/app/code/Magento/Quote/Model/Quote/Address/Total.php:116

then as V015 [public] x4 Method has been added

**Review by aleron75 — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @danielrussob, thank you for the review.
[ENGCOM-8500](https://jira.corp.magento.com/browse/ENGCOM-8500) has been created to process this Pull Request
:eight_spoked_asterisk: @danielrussob, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA Passed

### Manual testing scenario
1. From Admin Panel, create a Simple and a Virtual products
2. In _Stores - Configuration - Sales- Shipping Methods_, configure **FedEx** shipping method
3. Go to Storefront and Sign In as Customer with a valid address
4. Add the previously created Simple Product to Cart
5. Click on the mini cart - **View and Edit Cart
6. Dropdown **Estimate Shipping and Tax**
7. Observe the _Federal Express_ shipping prices
before_without_virtual
8. Add the previously created Virtual Product to Cart
9. Observe the _Federal Express_ shipping prices

**BEFORE** applying changes provided in the PR, the price has been changed as one more product has been added to the Cart
before_with_virtual

**AFTER** switching to the PR, the shipping price is not changed when a Virtual Product is added to Cart
after_without_virtual
after_with_virtual

Also tested with Downloadable product. The result is Expected, shipping price did not change after adding Downloadable product to the Cart

**Comment by engcom-Charlie** [comment]
Needs to approve since **SVC** falls with errors:

- `virtual_amount` and `base_virtual_amount` columns were added to `quote_address` table
- `setVirtualAmount`, `getVirtualAmount`, `setBaseVirtualAmount`, `getBaseVirtualAmount` public methods were added to `Magento\Quote\Model\Quote\Address\Total`

**Comment by engcom-Alfa** [comment]
Hi @sunilit42, @aleron75 & @danielrussob thanks for your contribution and collaboration! 

Since the ticket is older and was pending for approval from many days, before taking it thru the pending approval process, I tried to reproduce the issue and issue is reproducible in clean Magento `2.4-develop` instance as shown in below reproducible steps:

1. Have a FedEx shipping methods configured successfully and working fine.
2. Make sure the store location is set to United State (just to prevent hassled execution).
3. Make sure to have one simple product and one virtual product at least with some price amount greater than $500 (not mandatory to have with same price, but it is for easier calculation).
4. Login to front end with US shipping address customer.
5. Add the simple product to the cart and open view/edit cart details page.
6. Expand the Estimated Shipping and Tax section, and make sure to have the Country as United states and state with California and the valid pin code too.
7. Check the calculated price in the **_Federal Express_** shipping charges radio button section. (make a note of the value)
8. Add a virtual product as well into the cart, and again check the Fedex value in the same section.

There is a price increment in the FedEx shipping charges noticeable as shown in the below screenshots.

image

image

Now, the ticket must get updated with all merge conflicts sorted out!
Can @sunilit42 please update the code and revert us back, so that I can follow up on it later and process till the last stage!

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Because of the [above-comment](https://github.com/magento/magento2/pull/30434#issuecomment-1096687815), moving it to _changes Requested_.
Thanks

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
Hi @engcom-Alfa,

I have merged latest 2.4-develop and resolved all conflicts. Hence moving it further for testing.

Thank you!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento give me 2.3.5-p1 instance

**Comment by engcom-Lima** [comment]
@magento give me 2.3.5-p1 instance

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento give me 2.3.5-p1 instance

**Comment by engcom-Alfa** [comment]
@magento run all tests

**Comment by sunilit42** [comment]
do i need to any thing here?

**Comment by engcom-Alfa** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
> do i need to any thing here?

Hi @sunilit42 Thanks for asking!
Actually we are trying to reproduce the issue again, not finding the issue with latest `2.4-develop` code. Might have fixed with any other fixes implicitly.
Can you please help us instructing how to reproduce it exactly.
As off now, we tried the same way as tried in the last month [mentioned above](https://github.com/magento/magento2/pull/30434#issuecomment-1096687815), but we are unable to see the issue this time with the updated code. Could you please reproduce it with `2.4-develop` code once and give us clear steps to reproduce please, including if any configurations need to be done. If not reproducible, then we can close this ticket too!
Thanks in advance

**Comment by sunilit42** [comment]
@engcom-Alfa how i can get access of code or mysql ?

**Comment by engcom-Alfa** [comment]
> @engcom-Alfa how i can get access of code or mysql ?

Well! You have your code available in your branch here - https://github.com/sunilit42/magento2/tree/magento-issue-28748 , correct! So you just have to do the setup Magento instance in your local system with your local mysql database. You will be anyhow able to access your code and mysql.

If not, as we want to know if you can reproduce the issue, you may also follow the instructions give in [bot-comment](https://github.com/magento/magento2/pull/30434#issuecomment-706534603) and try to reproduce with Magento auto generated cloud instance too.
If it is not reproducible anymore, we will keep this ticket closed as it is!
Thanks :)

**Pull request closed without merging**
